test/integration/examples/apiserver_test.go

```diff
@@ -178,30 +183,32 @@ func TestAggregatedAPIServer(t *testing.T) {
 	// start the wardle server to prove we can aggregate it
 	go func() {
 		for {
-			// always get a fresh port in case something claimed the old one
-			wardlePortInt, err := framework.FindFreeLocalPort()
+			listener, port, err := genericapiserveroptions.CreateListener("tcp", "127.0.0.1:0")
 			if err != nil {
 				t.Fatal(err)
 			}
-			atomic.StoreInt32(wardlePort, int32(wardlePortInt))
-			wardleCmd := sampleserver.NewCommandStartWardleServer(os.Stdout, os.Stderr, stopCh)
-			wardleCmd.SetArgs([]string{
-				"--bind-address", "127.0.0.1",
-				"--secure-port", strconv.Itoa(wardlePortInt),
-				"--requestheader-username-headers=X-Remote-User",
-				"--requestheader-group-headers=X-Remote-Group",
-				"--requestheader-extra-headers-prefix=X-Remote-Extra-",
-				"--requestheader-client-ca-file=" + proxyCACertFile.Name(),
-				"--requestheader-allowed-names=kube-aggregator",
-				"--authentication-kubeconfig", kubeconfigFile.Name(),
-				"--authorization-kubeconfig", kubeconfigFile.Name(),
-				"--etcd-servers", framework.GetEtcdURL(),
-				"--cert-dir", wardleCertDir,
-				"--kubeconfig", kubeconfigFile.Name(),
-			})
-			if err := wardleCmd.Execute(); err != nil {
-				t.Log(err)
+			atomic.StoreInt32(wardlePort, int32(port))
+			options := sampleserver.NewWardleServerOptions(os.Stdout, os.Stderr)
+			options.Complete()
+
+			options.RecommendedOptions.SecureServing.BindAddress = net.ParseIP("127.0.0.1")
```

Comment:
this is a step back. The flags are more abstract. Don't like to switch back to options. Can we come up with an hybrid of flags plus the listener in the options?

Reply:
Maybe write a fake `NewCommandStartWardleServer` here

Reply:
Can't we create the options, parse the flags and patch in the listener? (have not looked at the code whether it's feasible) The goal is to have as few custom plumbing code in the test as possible. I.e. as end-2-end as possible.

Reply:
I tried but failed, because in  `NewCommandStartWardleServer` flags is parsed and conflicts with flags parse in test. And BTW we have to create options and parse flags , overwrite cobra.Command.RunE. This way equals to rewrite a `NewCommandStartWardleServer`.